File: practice_in_python/LRUCache.py
from collections import OrderedDict


class LRUCache:

    def __init__(self, capacity: int):

        self.size = capacity

        self.lru_cache = OrderedDict()

    def get(self, key: int) -> int:

        if key not in self.lru_cache:
            return -1

        else:
            # refresh the entry with given key
            self.lru_cache.move_to_end(key)

            return self.lru_cache[key]

    def put(self, key: int, value: int) -> None:

        if key not in self.lru_cache:

            if len(self.lru_cache) >= self.size:
                # pop the least used entry
                self.lru_cache.popitem(last=False)

        else:
            # refresh the entry with given key
            self.lru_cache.move_to_end(key)

        self.lru_cache[key] = value

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: practice_in_C++/data_structures/lru_cache.cpp
#include <bits/stdc++.h>

class LRUCache
{
private:
    std::unordered_map<int, std::list<std::pair<int, int>>::iterator> m;
    // doubly linked list
    std::list<std::pair<int, int>> l;
    int cap;

    // tail insertion
    void MoveToMostRecentlyUsed(int key)
    {
        // where, which list, and what value
        l.splice(l.end(), l, m[key]);
    }

public:
    LRUCache(int capacity)
    {
        cap = capacity;
    }
    ~LRUCache() {}

    int get(int key)
    {
        // couldn't find the key
        if (m.find(key) == m.end())
        {
            return -1;
        }

        MoveToMostRecentlyUsed(key);

        return m[key]->second;
    }

    void put(int key, int value)
    {
        if (m.find(key) != m.end())
        {
            m[key]->second = value;
            MoveToMostRecentlyUsed(key);
        }
        else
        {
            if (m.size() >= cap)
            {
                m.erase(l.front().first);
                l.pop_front();
            }

            l.push_back({key, value});
            auto tmpIter = l.end();
            // l.end() is pointing at the position
            // right after the last element
            // in the list, so decrement it to access the last element
            tmpIter--;
            m[key] = tmpIter;
        }
    }

    void printll()
    {
        // linked list contents
        for (const auto &node : l)
        {
            std::cout << node.first << ', ' << node.second << "\n";
        }
    }
};

/**
 * LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */